### PR TITLE
Parallel Cross Validation

### PR DESCRIPTION
**Pull request by TomAugspurger**
This changes how parallel cross validation works. It's an alternative to https://github.com/facebook/prophet/pull/1356. I think that a release hasn't been made with #1356, so hopefully this isn't breaking an existing API.

My primary motivation is to allow distributing the work on a cluster with Dask. As a precursor, https://github.com/facebook/prophet/commit/ed3639cfa204728486629f71b1c7ae025bf2c387 refactors things to use `concurrent.futures` for either thread- or process-based parallelism. Users choose parallelism with `parallel=None / "processes" / "threads"`.

https://github.com/facebook/prophet/commit/ed3639cfa204728486629f71b1c7ae025bf2c387 adds in Dask support. It provides a similar API to `concurrent.futures`, with a couple differences I'll annotate inline.

cc @ryankarlos (from #1356) if you have thoughts and @bletham.

---

Some rough timings

```
# parallel=None
CPU times: user 3min 17s, sys: 33.5 s, total: 3min 51s
Wall time: 43.1 s

parallel="threads"
CPU times: user 1min 5s, sys: 3min 27s, total: 4min 33s
Wall time: 45.4 s

# parallel="processes"
CPU times: user 1.18 s, sys: 318 ms, total: 1.49 s
Wall time: 12.2 s

# parallel="dask"
CPU times: user 954 ms, sys: 159 ms, total: 1.11 s
Wall time: 13.3 s
```

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -5,7 +5,7 @@ python:
 
 install:
   - pip install --upgrade pip
-  - pip install -U -r python/requirements.txt
+  - pip install -U -r python/requirements.txt dask[dataframe] distributed
```
Comment (TomAugspurger) [inline .travis.yml:8]: I suspect you don't want Dask as a required dependency, so I didn't add it to requirements.txt.

File: python/fbprophet/diagnostics.py [hunk 2]
```diff
@@ -82,9 +83,19 @@ def cross_validation(model, horizon, period=None, initial=None, multiprocess=Fal
...
-        core.
+    parallel : {None, 'processes', 'threads'}
+
+        How to parallelize the forecast computation. By default no parallelism
+        is used.
+
+        * None : No parallelism.
+        * 'processes' : Parallelize with concurrent.futures.ProcessPoolExectuor.
+        * 'threads' : Parallelize with concurrent.futures.ThreadPoolExecutor.
+            Note that some operations currently hold Python's Global Interpreter
```
Comment (TomAugspurger) [inline python/fbprophet/diagnostics.py:94]: This may be worth looking into. In theory, thread-based parallelism can be faster since you don't need to serialize the (largeish) DataFrames between processes. But it seems like a decent amount of operations in `single_cutoff_forecast` and what calls seems to hold the GIL.

I'm not sure how feasible it is to improve this. Things like manipulating a DataFrame (slicing / adding columns, etc.) hold the GIL. Things like linear algebra don't.

File: python/fbprophet/diagnostics.py [hunk 3]
```diff
@@ -127,11 +138,39 @@ def cross_validation(model, horizon, period=None, initial=None, multiprocess=Fal
...
+            itertools.cycle([model]),
+            cutoffs,
+            itertools.cycle([horizon]),
+            itertools.cycle([predict_columns]),
+        ]
+
+        predicts = pool.map(single_cutoff_forecast, *iterables)
+        if parallel == "dask":
+            # convert Futures to DataFrames
+            predicts = pool.gather(predicts)
```
Comment (TomAugspurger) [inline python/fbprophet/diagnostics.py:172]: Dask's Client.map is a bit different. It returns a List[Future] objects, while concurrent.futures returns the a List of the concrete results.

So we need this extra `gather` here to convert from futures to concrete results (DataFrames in this case)

File: python/fbprophet/diagnostics.py [hunk 4]
```diff
@@ -127,11 +138,39 @@ def cross_validation(model, horizon, period=None, initial=None, multiprocess=Fal
...
+        elif parallel == "dask":
+            try:
+                import dask
+                from dask.distributed import get_client
+            except ImportError:
+                raise ImportError("parallel='dask' requies the optional "
+                                  "dependency dask.")
+            pool = get_client()
+            # delay df and model to avoid large objects in task graph.
+            df, model = pool.scatter([df, model])
```
Comment (TomAugspurger) [inline python/fbprophet/diagnostics.py:159]: These objects are a bit large. Doing this pre-scattering helps improve the performance for Dask, and avoids a warning bubbling up to the user about large objects in the task graph.

**Comment by TomAugspurger** [comment]
I'll look into the CI failure. It's passing for me locally with `python -m unittest fbprophet/tests/test_diagnostics.py`, but failing with `python setup.py test`.

**Comment by ryankarlos** [comment]
@TomAugspurger  this looks great ! Now the user has an option for  multi-thread, multiprocessing and on a dask cluster. Yes the multiprocess bit wasnt pushed to pypy yet - so should break anything but @bletham  can confirm. 

One small question - how do the timings vary - between using `parallel=processes` and `parallel=dask` for large number of cutoffs (say if the horizon and period were reduced) - i suppose thread based parallelism would  be slightly faster with smaller dataframes in this case ?
Also, do you think it would be useful for user to choose the number of chunks/chunk size - assuming at the moment the chunk size is determined automatically by number of cores in machine and available memory ?
Im not very familiar with dask distributed (ive only used dask df and dask delayed previously)  - i gather   the url for the performance dashboard is automatically generated for the user ?

File: notebooks/diagnostics.ipynb [hunk 5]
```diff
@@ -263,7 +263,22 @@
...
     "In R, the argument `units` must be a type accepted by `as.difftime`, which is weeks or shorter. In Python, the string for `initial`, `period`, and `horizon` should be in the format used by Pandas Timedelta, which accepts units of days or shorter.\n",
     "\n",
-    "Cross-validation can also be run in multiprocessing mode in Python, by setting the `multiprocess` argument to `True`\n",
+    "Cross-validation can also be run in parallel mode in Python, by setting specifying the `parallel` keyword. Three modes are supported\n",
+    "\n",
+    "* `parallel=\"processes\"`\n",
+    "* `parallel=\"threads\"`\n",
+    "* `parallel=\"dask\"`\n",
+    "\n",
+    "For problems that aren't too big, we recommend using `parallel=\"processes\"`. It will achieve the highest performance when the parallel cross validation can be done on a single machine. For large problems, a [Dask](https://dask.org) cluster can be used to do the cross validation on many machines.\n",
```
Comment (ryankarlos) [inline notebooks/diagnostics.ipynb:272]: nit - could we add this line in -  'You will need to install dask 2.14.0 or above separately, as it will not by default be installed with fbprophet'.

File: python/fbprophet/tests/test_diagnostics.py [hunk 6]
```diff
@@ -44,7 +44,7 @@ def test_cross_validation(self):
 
         try:
             from dask.distributed import Client
-            client = Client()
+            client = Client(processes=False)  # noqa
```
Comment (ryankarlos) [inline python/fbprophet/tests/test_diagnostics.py:47]: Is this an issue just with `setup.py test` or does it also impact if the user runs it with `processes=True` - assuming the timings you listed in  are for multiple processes with single threaded workers ?

**Comment by TomAugspurger** [comment]
> Few questions - how do the timings vary...

I'll get some additional timings later today, including some on a real cluster. But the high-level overview is that

1. *Currently* parallel="threads"` will be slower than `parallel="processes"` in essentially every case. In theory thread-based parallelism should be faster when
  - a: The time to serialize data between processes is a large portion of the runtime. So something like `# of forecasts == # of CPU cores`, and a somewhat large DataFrame?
  - b: The GIL is not held too long inside `single_cutoff_forecast`.
2. Dask has no real benefit on a single-machine. It's really only useful on a cluster
3. Dask will be faster when the number of forecasts is much larger than the number of (local) CPU cores. Then the cost of serializing the data will be offset by the increased parallelism.


I've spent a bit of time profiling just what happens in `single_cutoff_forecast`. Nothing concrete yet though. It seems like quite a bit of time is spent in pandas, when we really just need NumPy arrays, but I need to do a bit more digging. Since NumPy generally releases the GIL, we might be able to improve the `threads` case (probably as a followup PR if that's OK).

> do you think it would be useful to provide an option to the user for choosing the number of chunks/chunk size

Yes, that could be useful. What do you think about accepting an initialized Pool?

```python
pool = concurrent.futures.ProcessPoolExecutor(4)
df_cv = cross_validation(..., pool=pool)
```

> but i gather that the url for the performance dashboard is automatically generated for the user ?

Yes. I'm going to get an example at `examples.dask.org`. I have one at https://mybinder.org/v2/gh/TomAugspurger/dask-examples/prophet?filepath=applications%2Fforecasting-with-prophet.ipynb. It's not using the new API proposed here yet, but things should look similar.

File: python/fbprophet/tests/test_diagnostics.py (hunk 6 above)
Comment (TomAugspurger) [reply python/fbprophet/tests/test_diagnostics.py:47]: Yes, it's only failing with `setup.py test`. For context, this typically happens when have a callable that creates processes at the top-level of a file.

```
# file.py
pool = multiprocess.Pool()  # or dask.distributed.Client()
```

I don't really see why it would be happening in a test function.

And those timings were all for one-thread per worker.

**Comment by ryankarlos** [comment]
> Since NumPy generally releases the GIL, we might be able to improve the `threads` case (probably as a followup PR if that's OK).

That sounds reasonable - i may have some time later next week once you have got this PR in, to modify `single_cutoff_forecast` to use primarily Numpy to see if we get any improvement for threads case  - as this is some of the more interesting things I get to do these days. 

> Yes, that could be useful. What do you think about accepting an initialized Pool?
> 
> ```python
> pool = concurrent.futures.ProcessPoolExecutor(4)
> df_cv = cross_validation(..., pool=pool)
> ```

yeh that sounds good -  guess the `parallel` keyword would be redundant then ? 

Im also working on running some HP tuning with this - and maybe can add that to the 
docs as well. Will post some stuff later.

**Comment by TomAugspurger** [comment]
On the topic of accepting callables for custom parallelism

> [@blethem] I do think it'd be good to allow for callables so that people in the future could try new functions without having to push them into the package first, but for the built-in ones having a string alias seems good.

> [@ryankarlos] yeh that sounds good - guess the parallel keyword would be redundant then ?

What do you think about accepting anything with a `.map(Callable, *iterables)`? This isn't quite as simple as just accepting a callable, but I think it's not too bad. I'm trying out an example with a [joblib's](https://joblib.readthedocs.io/en/latest/) loky backend now:

File: python/fbprophet/diagnostics.py [hunk 7]
```diff
@@ -127,11 +138,39 @@ def cross_validation(model, horizon, period=None, initial=None, multiprocess=Fal
...
+
+        iterables = [
+            itertools.cycle([df]),
+            itertools.cycle([model]),
+            cutoffs,
+            itertools.cycle([horizon]),
+            itertools.cycle([predict_columns]),
+        ]
```
[the remark this answers is not in the document]
Comment (TomAugspurger) [reply python/fbprophet/diagnostics.py]: Mmm that doesn't quite work for the `concurrent.futures.map`-style interface. It expects its `*iterables*` to be "grouped" by argument, rather than unpacked.

```python
iterables = [[a1_v1, a1_v2, a1_v3], [a2_v1, a2_v2, a2_v3]]  # want
iterables [(a1_v1, a2_v2), (a1_v2, a2_v2), (a1_v3, a2_v3)]  # unpacked
```

I've pushed up a change that's maybe a bit more readable, assuming you're familiar with how `zip(*iterables)` works.

**Comment by ryankarlos** [comment]
@bletham  @TomAugspurger  ive just experimented with some hyperparameter tuning as an example for users to implement in the docs - using normal for loop and dask.  Like to know your opinions - and how/if it would be useful to go on top of the parallel cross validation being implemented here. (im limited by a not so powerful machine so timings may seem slow but i think with a bit of optimisation in the code here we could get significant speedup espcially trying out a larger number of param combinations)

```
def single_param_cv(history_df, metrics, param_dict):
    m = fbprophet.Prophet(**param_dict)
    m.fit(history_df)
    df_cv = cross_validation(m, initial='2600 days', period='100 days', horizon = '200 days')
    df_p = performance_metrics(df_cv, rolling_window=1)
    df_p['params'] = str(param_dict)
    df_p = df_p.loc[:, metrics]
    return df_p

def param_grid_to_df(**param_dict):
    param_iter = itertools.product(*param_dict.values())
    params =[]
    for param in param_iter:
        params.append(param) 
    params_df = pd.DataFrame(params, columns=list(param_dict.keys()))
    return params_df


def hyperparameter_cv(history_df, params_df, single_cv_callable, pool):
    results = []
    for param in params_df.values:
        param_dict = dict(zip(params_df.keys(), param))
        if pool is None:
            predict = single_cv_callable(history_df, param_dict=param_dict)
            results.append(predict)
        elif isinstance(pool, dask.distributed.client.Client):
            remote_df = pool.scatter(history_df)
            future = pool.submit(single_cv_callable, remote_df, param_dict=param_dict)
            results.append(future)
        else:
            raise ValueError(f'Pool needs to be an instantiated dask distributed client object or None')
    if isinstance(pool, dask.distributed.client.Client):
        results = pool.gather(results)
    results_df = pd.concat(results)
    
    return results_df
```
**Parameter grid settings**
```
param_dict = {  
                'changepoint_prior_scale': [0.1, 1, 10],
                'changepoint_range': [0.8, 0.9]
              }

metrics = ['horizon', 'rmse', 'mape', 'params'] 

params_df = param_grid_to_df(**param_dict)
single_cv_callable = functools.partial(single_param_cv, metrics=metrics)
```
**Dask Distributed Futures API**
```
pool = Client()
pool

%time result_df  = hyperparameter_cv(df, params_df, single_cv_callable, pool=pool)
result_df
best_param = result_df.loc[result_df['rmse'] == min(result_df['rmse']), ['params']]
print(f'The best param combination is {best_param.values[0][0]}')
result_df
```
*Output*

The best param combination is {'changepoint_prior_scale': 0.1, 'changepoint_range': 0.8}
CPU times: user 10.8 s, sys: 1.41 s, total: 12.2 s
Wall time: 1min 36s

horizon | rmse | mape | params
-- | -- | -- | --
200 days | 0.453529 | 0.034922 | {'changepoint_prior_scale': 0.1, 'changepoint_...
200 days | 0.464545 | 0.036444 | {'changepoint_prior_scale': 0.1, 'changepoint_...
200 days | 0.460679 | 0.035948 | {'changepoint_prior_scale': 1.0, 'changepoint_...
200 days | 0.503636 | 0.040299 | {'changepoint_prior_scale': 1.0, 'changepoint_...
200 days | 0.463724 | 0.036518 | {'changepoint_prior_scale': 10.0, 'changepoint...
200 days | 0.505390 | 0.039331 | {'changepoint_prior_scale': 10.0, 'changepoint...

**Normal for loop**

```
%time result_df  = hyperparameter_cv(df, params_df, single_cv_callable, pool=None)
best_param = result_df.loc[result_df['rmse'] == min(result_df['rmse']), ['params']]
print(f'The best param combination is {best_param.values[0][0]}')
result_df
```
*Output*

The best param combination is {'changepoint_prior_scale': 0.1, 'changepoint_range': 0.8}
CPU times: user 2min 14s, sys: 6.43 s, total: 2min 21s
Wall time: 2min 25s

horizon | rmse | mape | params
-- | -- | -- | --
200 days | 0.453529 | 0.034922 | {'changepoint_prior_scale': 0.1, 'changepoint_...
200 days | 0.464545 | 0.036444 | {'changepoint_prior_scale': 0.1, 'changepoint_...
200 days | 0.460679 | 0.035948 | {'changepoint_prior_scale': 1.0, 'changepoint_...
200 days | 0.503636 | 0.040299 | {'changepoint_prior_scale': 1.0, 'changepoint_...
200 days | 0.463724 | 0.036518 | {'changepoint_prior_scale': 10.0, 'changepoint...
200 days | 0.505390 | 0.039331 | {'changepoint_prior_scale': 10.0, 'changepoint...

**Comment by ryankarlos** [comment]
Also the SearchCV in Dask mirrors scikit learn API for cross sectional data so couldnt really diretcly use those methods above https://ml.dask.org/hyper-parameter-search.html

**Comment by TomAugspurger** [comment]
https://github.com/facebook/prophet/issues/549 would be neat!

Those look reasonable. I'll try out your example with more parameters on https://binder.pangeo.io/.

**Comment by ryankarlos** [comment]
> #549 would be neat!
> 
> Those look reasonable. I'll try out your example with more parameters on https://binder.pangeo.io/.

Cool ! Other param settings would be  'seasonality_prior_scale':[0.1, 1, 10.0] and  'holidays_prior_scale':[0.1, 1, 10.0]. @bletham has listed a useful set of recommendations here https://github.com/facebook/prophet/issues/1381#issuecomment-597301902

**Comment by TomAugspurger** [comment]
Just ran those CV snippets on a grid with 108 hyper-parameter combinations. Each single one took ~30 - 60s, depending on the params.

I ran it on a cluster with 40 workers, one core each. It finished in 3.5 minutes (compared with 54 minutes, assuming 30s hyperparam combination). So the scaling seems to be pretty good.

https://nbviewer.jupyter.org/gist/TomAugspurger/8619e223ea42c922cd655d18d6c96424 has the rough notebook.

**Comment by ryankarlos** [comment]
> Just ran those CV snippets on a grid with 108 hyper-parameter combinations. Each single one took ~30 - 60s, depending on the params.
> 
> I ran it on a cluster with 40 workers, one core each. It finished in 3.5 minutes (compared with 54 minutes, assuming 30s hyperparam combination). So the scaling seems to be pretty good.
> 
> https://nbviewer.jupyter.org/gist/TomAugspurger/8619e223ea42c922cd655d18d6c96424 has the rough notebook.

oh wow nice - thats some massive time improvement !  You OK if I submit this as a separate PR for the docs once you get this PR in ? 
One more thing - is there any point in using dask delayed as an example as well - ive found it to be slightly slower https://examples.dask.org/applications/embarrassingly-parallel.html

**Comment by TomAugspurger** [comment]
Yeah, I think keeping the hyperparameter stuff as a followup makes sense.

When you used `dask.delayed` did you also have a distributed `Client()`
active? If not, `delayed` uses the `threaded` scheduler by default which
will suffer from GIL issues.

Otherwise I'm not sure why it would be slower. It shouldn't have much
overhead over the `Client.submit` style, especially since these are such
long-running tasks (10s of seconds).

**Comment by ryankarlos** [comment]
> When you used `dask.delayed` did you also have a distributed `Client()` active? If not, `delayed` uses the `threaded` scheduler by default which will suffer from GIL issues. 

Ahh yes - i had run it before `Client()` was active - so thats probably why !

**Comment by bletham** [comment]
This is great work, thanks for adding it in!

**Comment by TomAugspurger** [comment]
Thanks for the help! Please do ping me when people report issues :)
